src/osd/OSD.cc

```diff
@@ -5846,10 +5846,20 @@ void OSD::handle_scrub(MOSDScrub *m)
       PG *pg = p->second;
       pg->lock();
       if (pg->is_primary()) {
+        if (m->stop_scrub && !pg->scrubber.must_scrub) {
```

Comment:
See my comments below.  Maybe the stop_scrub code should all be handled at the top here and only get into the existing code when not stopping.